File: src/WordPattern.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class WordPattern {
    public static boolean wordPattern(String pattern, String string) {
        Map<Character, String> bijection = new HashMap<>();
        Set<String> wordsMappedTo = new HashSet<>();
        String[] words = string.split(" ");
        if (words.length != pattern.length()) {
            return false;
        }

        for (int index = 0 ; index < pattern.length() ; index++) {
            char letter = pattern.charAt(index);
            String word = words[index];

            if (!bijection.getOrDefault(letter, word).equals(word) ||
                    (!bijection.containsKey(letter) && wordsMappedTo.contains(word))) {
                return false;
            }

            bijection.put(letter, word);
            wordsMappedTo.add(word);
        }

        return true;
    }
}

File: python/word_pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split()
        if len(words) != len(pattern):
            return False
        mapping = {}
        used_words = set()
        for character, word in zip(pattern, words):
            if character in mapping:
                if mapping[character] != word:
                    return False
            elif word in used_words:
                return False
            else:
                mapping[character] = word
                used_words.add(word)
        return True


print(Solution().wordPattern('abba', 'dog dog dog dog'))
